sips/sip-026/sip-026-sbtc_peg.md

```diff
@@ -0,0 +1,171 @@
+
+## Comparison to Other Protocols
+
+### [WBTC](https://wbtc.network/assets/wrapped-tokens-whitepaper.pdf)
+WBTC is made up of 50+ merchants and custodians with keys to the WBTC multisig contract on Ethereum. WBTC deposits and withdrawals can only be performed by the authorized merchants and end users purchase WBTC directly from the merchants. Although the merchants manage issuance and redemption, all BTC backing WBTC is held by a single custodian.
+
+### [tBTC v2](https://whitepaper.io/document/691/tbtc-whitepaper)
+tBTC is an open membership system, where the BTC is managed by a rotating set of randomly selected nodes which manage a threshold wallet. The system requires that  51-of-100 randomly selected wallet signers must collaborate to produce a proper signature. 
+
+### [RBTC](https://rootstock.io/static/a79b27d4889409602174df4710102056/RS-whitepaper.pdf)
+Rootstock’s (RSK) 2-way peg protocol is called “the Powpeg”. Peg operations settle to Bitcoin via merge mining on the RSK side-chain. Peg operators are incentivized by earning a portion of Rootstock transaction fees. PowPeg operators keep specialized hardware called PowHSMs active and connected to special types of Rootstock full nodes. Since the Bitcoin blockchain and the Rootstock sidechain are not entangled in a single blockchain or in a parent-child relation, peg-in and peg-out transactions require a high number of block confirmations. Peg-ins require 100 Bitcoin blocks, and peg-outs require 200 Bitcoin blocks.
+
+## Activation
+sBTC Bootstrap is designed to activate on Stacks Nakamoto as defined in [SIP-021](https://github.com/stacksgov/sips/blob/feat/sip-021-nakamoto/sips/sip-021/sip-021-nakamoto.md). Therefore, this SIP is only meaningful when SIP-021 activates. The sBTC Working Group plans to observe at least 2-4 weeks of network behavior on Stacks Nakamoto to ensure a stable release. After this period, sBTC Bootstrap can be activated on the Stacks network without requiring a separate hard fork.
+
+### Process of Activation
+Users can vote to approve this SIP with either their locked/stacked STX or with unlocked/liquid STX, or both. The criteria for the stacker and non-stacker voting is as follows.
+
+**For Stackers:**
+In order for this SIP to activate, the following criteria must be met by the set of Stacked STX:
+- At least 80 million Stacked STX must vote at all to activate this SIP.
+- Of the Stacked STX that vote, at least 66% of them must vote "yes."
+
+The voting addresses will be;
+- **Bitcoin Yes Address:** 3Jq9UT81fnT2t24XjNVY7wijpsSmNSivbK
+- **Bitcoin No Address:** 3QGZ1fDa97yZCXpAnXQd6JHF4CBC6bk1r4
+- **Stacks Yes Address:** SP36GHEPEZPGD53G2F29P5NEY884DXQR7TX90QE3T
+- **Stacks No Address:** SP3YAKFMGWSSATYNCKXKJHE2Z5JJ6DH88E4T8XJPK
+
+which encode the hashes of the following phrases into bitcoin / stacks addresses:
+
+- Yes to A Decentralized Two-Way Bitcoin Peg
+- No to A Decentralized Two-Way Bitcoin Peg
```

Comment:
Describe how

Reply:
Following has been used @jcnelson where proposalId is the string to encode.
```
export async function generateAddresses(proposalId:string) {

  console.log('generateAddresses: ' + proposalId)
  let encoder = new TextEncoder();
  let encoded = encoder.encode(`Yes to ${proposalId}`);
  let hash256 = sha256(encoded);
  const hash160Y = ripemd160(hash256);

  encoder = new TextEncoder();
  encoded = encoder.encode(`No to ${proposalId}`);
  hash256 = sha256(encoded);
  const hash160N = ripemd160(hash256);

  const net = getNet(getConfig().network);
  
  const p2shObjY = btc.p2sh({type:'sh', script: btc.Script.encode(['DUP', 'HASH160', hash160Y, 'EQUALVERIFY', 'CHECKSIG'])}, net)
  const p2shObjN = btc.p2sh({type:'sh', script: btc.Script.encode(['DUP', 'HASH160', hash160N, 'EQUALVERIFY', 'CHECKSIG'])}, net)

  console.log(p2shObjY.address)
  console.log(p2shObjN.address)

  const netPrefix = (getConfig().network === 'testnet') ? 26 : 22;
  const yStxAddress = c32address(netPrefix, hex.encode(hash160Y))
  const nStxAddress = c32address(netPrefix, hex.encode(hash160N))

  const votingAddresses = {
    yBtcAddress: p2shObjY.address as string,
    nBtcAddress: p2shObjN.address as string,
    yStxAddress,
    nStxAddress,
  }
  console.log('generateAddresses: ', votingAddresses)
  return votingAddresses;
}
```